internal/pkg/agent/application/gateway/fleet/fleet_gateway.go

```diff
@@ -156,37 +181,158 @@ func (f *fleetGateway) Run(ctx context.Context) error {
 		case <-f.scheduler.WaitTick():
 			f.log.Debug("FleetGateway calling Checkin API")
 
-			// Execute the checkin call and for any errors returned by the fleet-server API
-			// the function will retry to communicate with fleet-server with an exponential delay and some
-			// jitter to help better distribute the load from a fleet of agents.
-			resp, err := f.doExecute(ctx, backoff)
+			checkinResponse, err := f.triggerCheckin(ctx, backoff)
 			if err != nil {
+				f.log.Errorf("triggering checkin failed: %v", err)
 				continue
 			}
 
-			actions := make([]fleetapi.Action, len(resp.Actions))
-			copy(actions, resp.Actions)
+			f.log.Debug("Processing checkin response")
+			actions := make([]fleetapi.Action, len(checkinResponse.Actions))
+			copy(actions, checkinResponse.Actions)
 			if len(actions) > 0 {
 				f.actionCh <- actions
 			}
 		}
 	}
 }
 
+func (f *fleetGateway) triggerCheckin(ctx context.Context, backoff backoff.Backoff) (*fleetapi.CheckinResponse, error) {
+	// get current state
+	stateUpdateSubCtx, cancelStateUpdateSub := context.WithCancel(ctx)
+	stateUpdateCh := f.stateFetcher.StateSubscribe(stateUpdateSubCtx).Ch()
+	defer func() {
+		cancelStateUpdateSub()
+	}()
+	state := <-stateUpdateCh
+
+	for {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+
+		checkinResult := f.performCancellableCheckin(ctx, state, stateUpdateCh, backoff, f.debouncerFactory())
+
+		if checkinResult.err != nil {
+			var relaunchCheckinErr *needNewCheckinError
+			if errors.As(checkinResult.err, &relaunchCheckinErr) {
+				// checkin was cancelled because we have an updated state to send to fleet
+				state = relaunchCheckinErr.newState
+			}
+
+			continue
+		}
+
+		// if we got here the checkin has been completed
+		return checkinResult.response, nil
+	}
+}
+
+func (f *fleetGateway) performCancellableCheckin(ctx context.Context, initialState state.State, stateUpdates <-chan state.State, backoff backoff.Backoff, dbouncer debouncer) checkinResult {
+	checkinID, _ := uuid.NewV4()
```

Comment:
This id associated to a checkin is there to be able to track what happens to checkin requests during the cancellatio/relaunch (since we now may have hanging cancellations). If we have a unique id generated from a span I am happy to use that as an identifier so that we can track it across the network boundary. I'll have a look and see if I can integrate it in this PR or if it takes too long I will create a separate issue for that

Reply:
Opened #2517